pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java

```diff
@@ -131,27 +139,75 @@ public String toExplainString() {
   }
 
   interface RangeEvaluator {
+
+    Set<FieldSpec.DataType> SUPPORTED_RAW_DATA_TYPES = EnumSet.of(FieldSpec.DataType.INT,
+            FieldSpec.DataType.LONG, FieldSpec.DataType.FLOAT, FieldSpec.DataType.DOUBLE);
+
+    static boolean canEvaluate(PredicateEvaluator predicateEvaluator, DataSource dataSource) {
+      if (dataSource.getRangeIndex() != null) {
+        boolean datatypeSupported = dataSource.getRangeIndex().isExact()
+                && (predicateEvaluator.isDictionaryBased()
+                || SUPPORTED_RAW_DATA_TYPES.contains(predicateEvaluator.getDataType()));
+        switch (predicateEvaluator.getPredicateType()) {
+          case EQ:
+            return datatypeSupported && dataSource.getRangeIndex().isExact()
+                    && predicateEvaluator instanceof EqualsPredicateEvaluatorFactory.EqualsPredicateEvaluator;
+          case RANGE:
+            return datatypeSupported && (predicateEvaluator instanceof SortedDictionaryBasedRangePredicateEvaluator
+                    || predicateEvaluator instanceof IntRawValueBasedRangePredicateEvaluator
+                    || predicateEvaluator instanceof LongRawValueBasedRangePredicateEvaluator
+                    || predicateEvaluator instanceof FloatRawValueBasedRangePredicateEvaluator
+                    || predicateEvaluator instanceof DoubleRawValueBasedRangePredicateEvaluator);
+          default:
+        }
+      }
+      return false;
+    }
+
     static RangeEvaluator of(RangeIndexReader<ImmutableRoaringBitmap> rangeIndexReader,
         PredicateEvaluator predicateEvaluator) {
-      if (predicateEvaluator instanceof SortedDictionaryBasedRangePredicateEvaluator) {
-        return new IntRangeEvaluator(rangeIndexReader,
-            (SortedDictionaryBasedRangePredicateEvaluator) predicateEvaluator);
+      if (predicateEvaluator.isDictionaryBased()) {
```

Comment:
Suggest branching on the predicate type first for readability. Basically add an if check on `predicateEvaluator.getPredicateType()` first, then do the switch

Reply:
Yes, I can do that. It's going to be a longer method and harder to fit on screen but I can see benefits.